Lib/posixpath.py

```diff
@@ -438,12 +418,15 @@ def _joinrealpath(path, rest, strict, seen):
             continue
         if name == pardir:
             # parent dir
-            if path:
-                path, name = split(path)
-                if name == pardir:
-                    path = join(path, pardir, pardir)
-            else:
+            if not path:
+                # ..
                 path = pardir
+            elif basename(path) == pardir:
+                # ../..
+                path = join(path, pardir)
+            else:
+                # foo/bar/.. -> foo
+                path = dirname(path)
```

Comment:
Is this OK? It was difficult to understand what the old code was doing.

Reply:
If it doesn't speed the code up, then no. If there's a _serious_ readability problem with this section of code, then we can _consider_ changing it in a different PR. But as we've said previously, anything that isn't related to performance improvements does _not_ belong in this PR. Otherwise it's highly confusing for people reading through the git history later to see a commit that says it's about optimising the code, but the commit actually changes loads of stuff that has nothing to do with performance improvements.

Reply:
It's pretty subtle, but I suspect your new code is a little slower.

The first case (`elif basename(path) == pardir:`) is reasonably rare - it means the last component of your in-progress resolved path is a `..`. This can only occur if the path begins with a `..` segment.

The latter case (`else:`) is more common - it means the last component isn't a `..` segment. In this case you're calling both `basename()` and `dirname()` in your new code, and that's _probably_ less performant than calling `split()` once.

Reply:
Maybe try measuring (1) `realpath('../..')` and (2) `realpath('foo/..')`. I suspect your patch makes (1) faster, but (2) slower. If that's the case, I think this change should be undone.

Reply:
You're absolutely right, I have modified the code to make `realpath('../..')` faster without making `realpath('foo/..')` slower.